묵찌빠 게임 [STEP 1] Toy, Charles

Charl-es opened the pull request:
@mint3382

안녕하세요 민트! 
STEP1 제출합니다.

---

### 고민했던 점
- 가위, 바위, 보를 어떤 타입으로 정의를 해야할 지 고민했었습니다.
    - 명확한 케이스가 존재하기 때문에 `enum` 타입을 사용했습니다.
- 가위 = 1, 바위 = 2, 보 = 3 이라는 값을 어떻게 표현을 할 지 고민했습니다.
   - `enum`타입에 `원시값`을 주어 각 `case`에 맞는 `원시값`을 부여하였습니다.
- 0부터 3이 아닌 값을 입력받은 경우에 분기처리 하는 방법에 대해 고민했습니다.
   - `if`문을 중첩시켜 바깥 `if`문을 `userChoice`가 0일 때와 0초과 4미만 일 때, 그리고 그 외의 경우로 분기 처리하였습니다.
      - 0일 때는 게임을 종료하도록 하였습니다.
      - 0 < `userChoice` < 4 일 때에는 컴퓨터 패와 사용자 패를 비교하여 승패 여부를 결정하는 분기 처리를 하였습니다.
      - 그 외의 경우는 잘못된 출력임을 나타내었고, 게임을 재시작 하도록 하였습니다.

---

### 조언을 얻고 싶은 점
- 리턴타입이 있는 함수에서 옵셔널 바인딩을 할 때 바인딩이 되지 않는 경우에 리턴 값을 어떤 식으로 작성해야 할 지 조언을 얻고 싶습니다.
```swift
private func inputUserNumber(number: String?) -> Int {
      guard let number = number, let number = Int(number) else { return 4 }
        
      return number
}
```

Charl-es committed:
Feat: 열거형, 구조체 정의

가위바위보의 정보를 담을 수 있는 enum 타입을 선언.
가위바위보의 게임을 관리하는 struct 타입 선언.

Developer-Groo committed:
Feat: 메뉴 출력 함수와 입력값 받는 함수

유저에게 보여줄 메뉴를 출력하는 함수를 구현
유저에게 받은 입력값의 옵셔널을 바인딩 하는 함수 구현

Charl-es committed:
Feat: playGame 함수 구현, 속성 추가

컴퓨터패와 유저의 패를 비교하여 결과값을 출력하는 함수 구현.
컴퓨터패와 유저패를 저장하는 속성 추가.

Developer-Groo committed:
Feat: 속성, 메서드 은닉화와 논리구조 수정

0~3이 아닌 값을 입력받은 경우의 로직 수정

Charl-es committed:
Feat: 불필요한 import와 조건문의 조건 수정

Developer-Groo committed:
Feat: 들여쓰기 제약사항 관련 수정

Charl-es committed:
Feat: 함수명과 아규먼트 레이블 수정

두 값을 비교하여 결과를 출력하는 함수명을 test123에서 compareValues로 변경.

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+//  RPSManger.swift
+//  RockPaperScissors
+//
+//  Created by 김형철 on 2023/09/05.
```
둘 중 누구의 이름이실까요? 🧐 ㅋㅋ 야곰이 만든 파일이 아니면 팀이 만들었다고 각자의 닉네임을 같이 적어주면 좋을 것 같아요~ 
함께 만든 프로젝트 파일이니까요! 😃

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+//  Created by 김형철 on 2023/09/05.
+//
+
+struct RPSManager {
```
가위바위보 매니저군요! 그런데 API Design Guideline에서는 코드를 읽는 사람의 모호함을 피하기 위해  모든 단어를 적어 달라 말하고 있습니다. 더 정확한 명칭을 사용해보는 건 어떨까요? 

[API Design Guidelines](https://www.swift.org/documentation/api-design-guidelines/)

Charl-es replied:
모호성 제거를 위하여 풀네임으로 수정하였습니다!

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+        guard let number = number, let number = Int(number) else { return 4 }
+        
+        return number
+    }
```
number라는 네이밍이 너무 중복되는 것 같아요! 
1. 바깥에서 이 함수를 읽었을 때 inputUserNumber(number:)가 되는데, argument label을 활용해보면 좋을 것 같아요. 
2. number는 타입이 Int여야 할것 같은데 왜 String인가요? 더 적합한 네이밍이 있지 않을까요? 🧐
3. 이 부분에 관해 남겨주셨던 조언과 관련한 부분은 코멘트에 있습니다!

Charl-es replied:
해당 부분 아규먼트 레이블 추가하여 네이밍 다시 하였습니다!

Charl-es replied:
이 부분에 대해서는 코멘트 주신대로 다른 `옵셔널 바인딩`을 사용해서 구현해 보았었습니다.
```swift
 private func inputUserNumber(number: String?) -> Int {
    if let number = number, let number = Int(number) {
        return number
    }

    return 4
 }
```
`if let` 경우에도 바인딩이 된다면 바인딩된 수를 리턴해주고, 바인딩이 안된다면 다른 반환값을 주어야하는 똑같은 상황이 발생했었는데, 이러한 반환값을 리턴해야 할 수 밖에 없는 상황에서는 어떤값을 리턴을 해야할 지 피드백을 얻고 싶습니다.

mint3382 replied:
반환값을 리턴 해야만 하는 상황인데 조건이 불충분할 수가 절대 없는 상황이라면, 그래서 어떤 값을 넣어도 상관 없는 경우라면 그 코드는 불필요한 코드가 아닐까요? 사용자에게 입력을 받는 부분을 따로 기능분리로 구현해보시는 건 어때요? 지금 readline()은 playGame()에 존재하는데 이를 검증하고 바인딩을 하는 건 inputUserNumber()입니다. 기능 분리의 단위에 대해 고민해보시는 건 어떨까요?

mint3382 commented on RockPaperScissors/RockPaperScissors/RockPaperScissors.swift:
```diff
@@ -0,0 +1,12 @@
+//  RockPaperScissors.swift
+//  RockPaperScissors
+//
+//  Created by 김형철 on 2023/09/05.
```
여기도 팀원 모두의 이름으로 바꿔주면 좋을 것 같아요! 🥰

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+        } else if userChoice > 0 && userChoice < 4 {
+            compareValues(computerCard, userCard)
+            return
+        } else {
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+            playGame()
```
playGame()에서 playGame()을 부르는 재귀의 형태를 띄고 있네요! 😀
1. 재귀가 아닌 방법으로는 무엇이 있을까요?
2. 둘의 장단점도 설명해주시면 좋겠습니다.

Charl-es replied:
1. 재귀함수
- 함수 자체를 기본적으로 호출합니다.
   - 함수가 호출될 때마다 스텍이 만들어 지는데 재귀함수의 경우 함수를 종료할 특정 조건을 주지 않는 다면, 스택 오버플로우를 일으킬 수 있습니다.
- 코드 길이와 변수가 적어 가독성이 높아집니다.
- 반복문 보다 비교적 속도가 느립니다.

2. 반복문
- 일련의 명령을 반복적으로 실행합니다.
- 속도가 빠릅니다.
- 재귀함수에 비해 코드의 가독성이 떨어집니다.

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+        return number
+    }
+    
+    private func printMenu() {
+        print("가위(1), 바위(2), 보(3)! <종료 : 0> :", terminator: " ")
+    }
```
처음 메뉴를 print하는 부분을 분리하여 주셨군요! 
함수가 하나의 기능만을 가지고 있을 수 있도록 하는 기능 분리가 잘 지켜진 것 같아요!
아주 좋아요~ 😍

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+            print("이겼습니다!")
+            return
+        }
+    }
```
1. if - else 문으로 작성해 주셨는데 switch도 사용해 보셨을까요?
2. print("졌습니다!")가 3번 반복되는데 반복되는 코드를 줄이는 방법이 있을까요?

Charl-es replied:
1. `if-else`문을 사용에 대한 이유
- `if-else`문의 경우, `Bool`타입의 결과 값에 따라 두 개의 흐름으로 갈라지는 분기처리 방법입니다.
- `switch`문의 경우는 표현식의 값에 따라 여러 값들과의 일치여부를 판단하는 분기처리 방법입니다.
- 저희의 코드 같은 경우에는 컴퓨터의 패와 사용자의 패를 비교하여 `Bool` 타입의 결과를 반환하도록 구현하였기 때문에 `if-else`문이 더 적합하다 생각되어 사용하였습니다.

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,60 @@
+        print("가위(1), 바위(2), 보(3)! <종료 : 0> :", terminator: " ")
+    }
+    
+    mutating func compareValues(_ computerValue: RockPaperScissors?, _ userValue: RockPaperScissors?) {
```
여기서는 argument label을 사용해주셨네요! 좋아요🥰 
그런데 이 메서드에는 왜 private를 붙이지 않으셨는지 알 수 있을까요?

Charl-es replied:
이 부분에 대해서는 새로운 함수를 구현하면서 누락되었던 것 같습니다 😅

mint3382 commented:
안녕하십니까, 토이 & 찰스! 민트입니다~😈
일주일 동안 잘 부탁드려요!

### Git commit convention
커밋 내역으로 설명을 적으신 부분이 대단합니다! 칭찬해요~ 😍
다만 수정하는 경우에도 feat: 이라고 적으셨던데 이유가 궁금합니다. 야곰 캠프 사이트 course에서 git commit 글을 읽어보시면 좋을 것 같아요.

### Git branch main
깃 브런치 분리가 되지 않고 메인에서 바로 작업하신 것 같은데 앞으로는 스텝 별로 브런치를 새로 파고 PR이 머지 되면 메인에 해당 스텝 브런치를 머지해주는 방식이 관리하기 좋을 것 같아요~ 

### Optional Binding 
옵셔널 바인딩을 하는 방법은 guard 문 외에도 존재합니다. 다른 방법들을 고려해보는 것이 좋을 것 같아요! 

[optional docs](https://developer.apple.com/documentation/swift/optional)

Developer-Groo committed:
Refactor: Pr 피드백 수정

-프로젝트 파일 이름 변경
-구조체 네이밍 변경
-inputUserNumber함수의 아규먼트레이블과 파라미터이름 변경
-compareValues함수안의 if문 조건 나열 방식 수정
-private키워드 추가

Charl-es committed:
Refactor: inputUserNumber 함수의 오류처리

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,65 @@
+        print("가위(1), 바위(2), 보(3)! <종료 : 0> :", terminator: " ")
+    }
+    
+    private mutating func compareValues(_ computerValue: RockPaperScissors?, _ userValue: RockPaperScissors?) {
```
이 부분 매개변수 수정을 잊으신 것 같아요! 다음 스텝에서 함께 진행해주세요~

mint3382 commented on RockPaperScissors/RockPaperScissors/RPSManager.swift:
```diff
@@ -0,0 +1,65 @@
+    mutating func playGame() {
+        printMenu()
+        
+        do {
+            let userChoice = try inputUserNumber()
+            
+            userCard = RockPaperScissors(rawValue: userChoice)
+            computerCard = RockPaperScissors(rawValue: Int.random(in: 1...3))
+            
+            if userChoice == 0 {
+                print("게임종료")
+                return
+            } else if userChoice > 0 && userChoice < 4 {
+                compareValues(computerCard, userCard)
+                return
+            } else {
+                print("잘못된 입력입니다. 다시 시도해주세요.")
+                playGame()
+            }
+        } catch InputError.wrongNumber {
+            print("잘못된 입력입니다. 다시 시도해주세요.")
+            playGame()
+        } catch {
+            print(error)
+        }
+    }
```
에러 타입을 사용하셨군요..! 시도해 보시는 것 아주 좋습니다.
다만 아까 디코에서 주셨던 질문과 연관 지어서 어차피 나올일 없는 에러라면, 저는 기능분리를 할 때 
- playGame()에서 readline() 상수를 입력받고,
- 그 상수가 잘못된 입력인지 아닌지 Bool값을 return 해주는 함수
- 해당 Bool 값에 따라 서로 다른 로직 처리를 해주는 함수
이렇게 구분해줄 것 같네요. 
야곰에서 항상 말하듯 코드에 정답은 없다고 생각합니다. 
머지해드릴테니 기능 분리에 대해서 한번 더 고민하고 다음 스텝에서 적용해보시면 좋을 것 같아요~😈
